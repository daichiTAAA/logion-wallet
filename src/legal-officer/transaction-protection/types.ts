import { UUID } from "../../logion-chain/UUID"

export type LocItemStatus = 'DRAFT' | 'PUBLISHED'

export type LocItemType = 'Data' | 'Document' | 'Linked LOC'

export interface LocItem {
    name: string,
    value: string,
    timestamp: string | null,
    type: LocItemType,
    submitter: string,
    status: LocItemStatus,
    nature?: string,
    target?: UUID,
    newItem: boolean,
}

export enum PublishStatus {
    NONE,
    START,
    PUBLISH_PENDING,
    PUBLISHING,
    PUBLISHED
}

export interface PublishState {
    status: PublishStatus;
}

export interface PublishProps {
    locItem: LocItem
}
