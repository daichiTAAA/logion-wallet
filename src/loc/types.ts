import { SignAndSubmit } from "../ExtrinsicSubmitter"
import { UUID } from "../logion-chain/UUID"

export type LocItemStatus = 'DRAFT' | 'PUBLISHED'

export type LocItemType = 'Data' | 'Document' | 'Linked LOC'

export interface LocItem {
    name: string,
    value: string,
    timestamp: string | null,
    type: LocItemType,
    submitter: string,
    status: LocItemStatus,
    nature?: string,
    target?: UUID,
    newItem: boolean,
    linkDetailsPath?: string,
}

export enum PublishStatus {
    NONE,
    START,
    PUBLISH_PENDING,
    PUBLISHING,
    PUBLISHED,
    ERROR
}

export interface PublishState {
    status: PublishStatus;
}

export interface PublishProps {
    locItem: LocItem;
    itemType: 'Public Data' | 'Document' | 'Link';
    signAndSubmitFactory: ((item: LocItem) => SignAndSubmit) | null;
    confirm: ((item: LocItem) => void) | null;
}

export type Viewer = 'User' | 'LegalOfficer';
