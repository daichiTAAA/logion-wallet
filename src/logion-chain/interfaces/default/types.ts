// Auto-generated via `yarn polkadot-types-from-defs`, do not edit
/* eslint-disable */

import type { Bytes, Enum, Option, Struct, Vec, bool, u128, u32, u64, u8 } from '@polkadot/types';
import type { AccountId, Balance, Hash, MultiAddress } from '@polkadot/types/interfaces/runtime';
import type { AccountInfoWithDualRefCount } from '@polkadot/types/interfaces/system';
import type { DepositBalance } from '@polkadot/types/interfaces/uniques';

/** @name AccountInfo */
export interface AccountInfo extends AccountInfoWithDualRefCount {}

/** @name Address */
export interface Address extends MultiAddress {}

/** @name AssetDetails */
export interface AssetDetails extends Struct {
  readonly owner: AccountId;
  readonly issuer: AccountId;
  readonly admin: AccountId;
  readonly freezer: AccountId;
  readonly supply: Balance;
  readonly deposit: DepositBalance;
  readonly max_zombies: u32;
  readonly min_balance: Balance;
  readonly zombies: u32;
  readonly accounts: u32;
  readonly is_frozen: bool;
}

/** @name AssetId */
export interface AssetId extends u64 {}

/** @name AssetMetadata */
export interface AssetMetadata extends Struct {
  readonly deposit: DepositBalance;
  readonly name: Bytes;
  readonly symbol: Bytes;
  readonly decimals: u8;
}

/** @name File */
export interface File extends Struct {
  readonly hash: Hash;
  readonly nature: Bytes;
}

/** @name LegalOfficerCaseOf */
export interface LegalOfficerCaseOf extends Struct {
  readonly owner: AccountId;
  readonly requester: AccountId;
  readonly metadata: Vec<MetadataItem>;
  readonly files: Vec<File>;
  readonly closed: bool;
  readonly loc_type: LocType;
  readonly links: Vec<LocLink>;
  readonly void_info: Option<LocVoidInfo>;
  readonly replacer_of: Option<LocId>;
}

/** @name LocId */
export interface LocId extends u128 {}

/** @name LocLink */
export interface LocLink extends Struct {
  readonly id: LocId;
  readonly nature: Bytes;
}

/** @name LocType */
export interface LocType extends Enum {
  readonly isTransaction: boolean;
  readonly isIdentity: boolean;
  readonly type: 'Transaction' | 'Identity';
}

/** @name LocVoidInfo */
export interface LocVoidInfo extends Struct {
  readonly replacer: Option<LocId>;
}

/** @name LookupSource */
export interface LookupSource extends MultiAddress {}

/** @name MetadataItem */
export interface MetadataItem extends Struct {
  readonly name: Bytes;
  readonly value: Bytes;
}

/** @name PeerId */
export interface PeerId extends Bytes {}

/** @name StorageVersion */
export interface StorageVersion extends Enum {
  readonly isV1: boolean;
  readonly isV2MakeLocVoid: boolean;
  readonly type: 'V1' | 'V2MakeLocVoid';
}

/** @name TAssetBalance */
export interface TAssetBalance extends u128 {}

export type PHANTOM_DEFAULT = 'default';
