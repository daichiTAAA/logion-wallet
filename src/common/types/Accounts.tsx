import moment, { Moment } from 'moment';
import { InjectedAccountWithMeta } from '@polkadot/extension-inject/types';

export interface Token {
    readonly value: string;
    readonly expirationDateTime: Moment;
}

export interface Account {
    readonly name: string,
    readonly address: string,
    readonly isLegalOfficer: boolean,
    readonly token?: Token,
}

export default interface Accounts {
    readonly all: Account[],
    readonly current?: Account,
}

export class AccountTokens {

    constructor(initialState: Record<string, Token>) {
        this.store = { ...initialState };
    }

    private store: Record<string, Token>;

    get(address: string): Token | undefined {
        return this.store[address];
    }

    merge(tokens: AccountTokens): AccountTokens {
        const newStore = { ...this.store };
        for(const address of tokens.addresses) {
            newStore[address] = tokens.store[address];
        }
        return new AccountTokens(newStore);
    }

    get addresses(): string[] {
        return Object.keys(this.store);
    }

    cleanUp(now: Moment): AccountTokens {
        const newStore: Record<string, Token> = {};
        for(const address of this.addresses) {
            const token = this.get(address)!;
            if(token.expirationDateTime.isAfter(now)) {
                newStore[address] = token;
            }
        }
        return new AccountTokens(newStore);
    }

    equals(other: AccountTokens): boolean {
        if(this.length !== other.length) {
            return false;
        }
        for(const address of this.addresses) {
            const thisToken = this.get(address);
            const otherToken = other.get(address);
            if(thisToken!.value !== otherToken?.value
                || !thisToken!.expirationDateTime.isSame(otherToken.expirationDateTime)) {
                return false;
            }
        }
        return true;
    }

    get length(): number {
        return this.addresses.length;
    }

    isAuthenticated(now: Moment, address: string | undefined): boolean {
        if(address === undefined) {
            return false;
        }
        const token = this.get(address);
        if(token === undefined) {
            return false;
        } else {
            return token.expirationDateTime.isAfter(now);
        }
    }

    earliestExpiration(): Moment | undefined {
        let earliest: Moment | undefined;
        for(const address of this.addresses) {
            const expiration = this.store[address].expirationDateTime;
            if(earliest === undefined || earliest.isAfter(expiration)) {
                earliest = expiration;
            }
        }
        return earliest;
    }
}

export function buildAccounts(
    injectedAccounts: InjectedAccountWithMeta[],
    userAddress: string | undefined,
    tokens: AccountTokens,
    isLegalOfficer: (address: string) => boolean,
): Accounts {
    const selectedAddress = currentOrDefaultAddress(injectedAccounts, userAddress, tokens);

    let current: Account | undefined;
    const matchingAddress = injectedAccounts.find(injectedAccount => injectedAccount.address === selectedAddress);
    if(selectedAddress !== undefined && matchingAddress !== undefined) {
        current = {
            name: matchingAddress.meta.name!,
            address: selectedAddress!,
            isLegalOfficer: isLegalOfficer(selectedAddress),
            token: tokenOrUndefinedIfExpired(tokens.get(selectedAddress)),
        }
    } else {
        current = undefined;
    }

    return {
        current,
        all: injectedAccounts.map(injectedAccount => ({
            name: injectedAccount.meta.name!,
            address: injectedAccount.address,
            isLegalOfficer: isLegalOfficer(injectedAccount.address),
            token: tokenOrUndefinedIfExpired(tokens.get(injectedAccount.address)),
        }))
    }
}

function currentOrDefaultAddress(
    injectedAccounts: InjectedAccountWithMeta[],
    currentAddress: string | undefined,
    tokens: AccountTokens
): string | undefined {
    const loggedAddresses = tokens.addresses;
    if(currentAddress !== undefined) {
        return currentAddress;
    } else if(loggedAddresses.length > 0) {
        const defaultAccount = injectedAccounts.find(account => loggedAddresses.includes(account.address));
        return defaultAccount?.address;
    } else {
        return undefined;
    }
}

function tokenOrUndefinedIfExpired(token: Token | undefined): Token | undefined {
    if(token === undefined) {
        return undefined;
    }

    const expirationDateTime = token.expirationDateTime;
    if(expirationDateTime.isBefore(moment())) {
        return undefined;
    } else {
        return token;
    }
}
