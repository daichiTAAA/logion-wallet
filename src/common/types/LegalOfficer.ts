export const DEFAULT_LEGAL_OFFICER = "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY"; // Alice
export const ANOTHER_LEGAL_OFFICER = "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty"; // Bob

const LEGAL_OFFICERS: string[] = [ DEFAULT_LEGAL_OFFICER, ANOTHER_LEGAL_OFFICER ];

export function isLegalOfficer(address: string): boolean {
    return LEGAL_OFFICERS.includes(address);
}

export interface LegalOfficer {
    name: string,
    address: string,
    details: string
}

function fakeAddress(i: number) {
    return "1234567890-abcd-" + i;
}

const defaultDetails = "Place de le République Française, 10";

export const legalOfficers: LegalOfficer[] = [
    {name: "Patrick Gielen", address: DEFAULT_LEGAL_OFFICER, details: "Modero Bruxelles\nCentral administration \nNijverheidslaan 1\nB - 1853 Grimbergen (Strombeek-Bever)\nBelgium"},
    {name: "Guillaume Grain", address: ANOTHER_LEGAL_OFFICER, details: "SELARL ADRASTEE\nGare des Brotteaux\n14, place Jules Ferry\n69006 LYON\nFrance"},
    {name: "Alain Barland", address: fakeAddress(5), details: defaultDetails},
    {name: "Ronald Vanswijgenhoven", address: fakeAddress(6), details: defaultDetails},
    {name: "Thibaut Barnier", address: fakeAddress(7), details: defaultDetails},
    {name: "Jona Van Leeuwen", address: fakeAddress(8), details: defaultDetails},
    {name: "Carlos Calvo", address: fakeAddress(9), details: defaultDetails},
    {name: "Ilias Tsipos", address: fakeAddress(10), details: defaultDetails},
    {name: "Pedroso Leal", address: fakeAddress(11), details: defaultDetails},
    {name: "Marc Schmitz", address: fakeAddress(12), details: defaultDetails},
];

export function getOfficer(address: string | undefined): LegalOfficer | null {
    if(address === null) {
        return null;
    }

    for(let i = 0; i < legalOfficers.length; ++i) {
        const legalOfficer = legalOfficers[i];
        if(legalOfficer.address === address) {
            return legalOfficer;
        }
    }
    return null;
}

export function legalOfficerByAddress(address: string): LegalOfficer {
    const legalOfficer = legalOfficers.find(legalOfficer => legalOfficer.address === address);
    if(legalOfficer !== undefined) {
        return legalOfficer;
    } else {
        throw new Error(`No legal officer with address ${address}`);
    }
}
