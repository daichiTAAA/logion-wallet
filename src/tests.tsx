import { ReactElement } from 'react';
import { createRenderer } from 'react-test-renderer/shallow';
import reactRenderer, { ReactTestRenderer, ReactTestRendererJSON, act as reactAct } from 'react-test-renderer';

export function render(element: ReactElement): null | ReactTestRendererJSON | ReactTestRendererJSON[] {
    const component = reactRenderer.create(element);
    return component.toJSON();
}

export function renderer(element: ReactElement): ReactTestRenderer {
    return reactRenderer.create(element);
}

export function shallowRender<E extends ReactElement>(element: ReactElement): E {
    const renderer = createRenderer();
    renderer.render(element);
    return renderer.getRenderOutput();
}

export function mockAccount(address: string, name: string) {
    return {
        address,
        meta: {
            name
        }
    };
}

export { reactAct as act };
