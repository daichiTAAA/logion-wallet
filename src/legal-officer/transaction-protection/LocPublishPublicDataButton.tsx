import { useLocContext } from "./LocContext";
import { LocItem } from "./types";
import LocPublishButton from "./LocPublishButton";

export interface Props {
    locItem: LocItem;
}

export default function LocPublishPublicDataButton(props: Props) {
    const { publishMetadata, confirmMetadata } = useLocContext();

    return (
        <LocPublishButton
            locItem={ props.locItem }
            confirm={ confirmMetadata }
            signAndSubmitFactory={ publishMetadata }
            itemType="Public Data"
        />
    )
}
