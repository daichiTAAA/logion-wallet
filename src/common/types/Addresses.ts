import moment, { Moment } from 'moment';
import { InjectedAccountWithMeta } from '@polkadot/extension-inject/types';
import { isLegalOfficer } from './LegalOfficer';

export interface Token {
    readonly value: string;
    readonly expirationDateTime: Moment;
}

export interface AccountAddress {
    readonly name: string,
    readonly address: string,
    readonly isLegalOfficer: boolean,
    readonly token?: Token,
}

export default interface Addresses {
    readonly addresses: AccountAddress[],
    readonly currentAddress?: AccountAddress,
}

export type AccountTokens = Record<string, Token>;

export function buildAddresses(
    injectedAccounts: InjectedAccountWithMeta[],
    userAddress: string | undefined,
    tokens: AccountTokens
): Addresses {
    const selectedAddress = currentOrDefaultAddress(injectedAccounts, userAddress, tokens);

    let currentAddress: AccountAddress | undefined;
    const matchingAddress = injectedAccounts.find(injectedAccount => injectedAccount.address === selectedAddress);
    if(selectedAddress !== undefined && matchingAddress !== undefined) {
        currentAddress = {
            name: matchingAddress.meta.name!,
            address: selectedAddress!,
            isLegalOfficer: isLegalOfficer(selectedAddress),
            token: tokenOrUndefinedIfExpired(tokens[selectedAddress]),
        }
    } else {
        currentAddress = undefined;
    }

    return {
        currentAddress,
        addresses: injectedAccounts.map(injectedAccount => ({
            name: injectedAccount.meta.name!,
            address: injectedAccount.address,
            isLegalOfficer: isLegalOfficer(injectedAccount.address),
            token: tokenOrUndefinedIfExpired(tokens[injectedAccount.address]),
        }))
    }
}

function currentOrDefaultAddress(
    injectedAccounts: InjectedAccountWithMeta[],
    currentAddress: string | undefined,
    tokens: AccountTokens
): string | undefined {
    const loggedAddresses = Object.keys(tokens);
    if(currentAddress !== undefined) {
        return currentAddress;
    } else if(loggedAddresses.length > 0) {
        const defaultAccount = injectedAccounts.find(account => loggedAddresses.includes(account.address));
        return defaultAccount?.address;
    } else {
        return undefined;
    }
}

function tokenOrUndefinedIfExpired(token: Token | undefined): Token | undefined {
    if(token === undefined) {
        return undefined;
    }

    const expirationDateTime = token.expirationDateTime;
    if(expirationDateTime.isBefore(moment())) {
        return undefined;
    } else {
        return token;
    }
}
