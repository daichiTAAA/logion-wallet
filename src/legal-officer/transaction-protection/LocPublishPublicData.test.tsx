jest.mock("./LocContext");

import { LocItem } from "./types";
import { shallowRender } from "../../tests";
import LocPublishPublicDataButton from "./LocPublishPublicDataButton";

describe("LocPublishPublicData", () => {

    const locItem:LocItem = {
        name: "data-name",
        type: "Data",
        status: "DRAFT",
        submitter: "data-submitter",
        value: "data-value",
        nature: "data-nature",
        timestamp: null
    }

    it("renders", () => {
        const tree = shallowRender(<LocPublishPublicDataButton locItem={ locItem } />)
        expect(tree).toMatchSnapshot();
    })
})
