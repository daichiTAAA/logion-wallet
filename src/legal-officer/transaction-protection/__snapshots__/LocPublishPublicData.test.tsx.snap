// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocPublishPublicData renders 1`] = `
<React.Fragment>
  <Button
    onClick={[Function]}
    variant="polkadot"
  >
    <Icon
      icon={
        Object {
          "id": "publish",
        }
      }
    />
     Publish to the blockchain
  </Button>
  <ProcessStep
    active={false}
    nextSteps={
      Array [
        Object {
          "buttonText": "Cancel",
          "buttonVariant": "secondary-polkadot",
          "callback": [Function],
          "id": "cancel",
          "mayProceed": true,
        },
        Object {
          "buttonText": "Publish",
          "buttonVariant": "polkadot",
          "callback": [Function],
          "id": "publish",
          "mayProceed": false,
        },
      ]
    }
    title="Publish Data (1/2)"
  >
    <Alert
      variant="info"
    >
      <p>
        Warning: after processing and blockchain publication, these data will be definitely and publicly available on the blockchain.
      </p>
    </Alert>
  </ProcessStep>
  <ProcessStep
    active={false}
    nextSteps={
      Array [
        Object {
          "buttonText": "OK",
          "buttonVariant": "primary",
          "callback": [Function],
          "id": "ok",
          "mayProceed": false,
        },
      ]
    }
    title="Publish Data (2/2)"
  >
    <ExtrinsicSubmitter
      id="publishMetadata"
      onError={[Function]}
      onSuccess={[Function]}
      signAndSubmit={null}
      successMessage="LOC public data successfully published"
    />
  </ProcessStep>
</React.Fragment>
`;
